lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.Common.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimePlatformAgnosticPch.h"
+#include "UnicodeText.h"
+
+namespace PlatformAgnostic
+{
+namespace UnicodeText
+{
+CharacterTypeFlags GetLegacyCharacterTypeFlags(char16 ch)
+{
+    Assert(ch >= 128);
+
+    const char16 lineSeperatorChar = 0x2028;
+    const char16 paraSeperatorChar = 0x2029;
+
+    if (ch == lineSeperatorChar || ch == paraSeperatorChar)
+    {
+        return LineCharGroup;
+    }
+
+    CharacterClassificationType charClass = GetLegacyCharacterClassificationType(ch);
+
+    if (charClass == CharacterClassificationType::Letter)
+    {
+        return LetterCharGroup;
+    }
+    else if (charClass == CharacterClassificationType::DigitOrPunct)
+    {
+        return IdChar;
+    }
+    else if (charClass == CharacterClassificationType::Whitespace)
+    {
+        return SpaceChar;
+    }
+
+    return UnknownChar;
+}
+
+#if !defined(_WIN32) || ENABLE_TEST_PLATFORM_AGNOSTIC
+namespace Internal
+{
+template <typename CharType>
+inline bool isDigit(__in CharType c)
+{
+    return c >= '0' && c <= '9';
+}
+
+template <typename CharType>
+inline int readNumber(__inout CharType* &str)
+{
+    int num = 0;
+
+    // Count digits on each string
+    while (isDigit(*str))
+    {
+        num = (num * 10) + (*str - '0');
+        str++;
+    }
+
+    return num;
+}
+
+///
+/// Implementation of CompareString(NORM_IGNORECASE | SORT_DIGITSASNUMBERS)
+/// This code is in the common library so that we can unit test it on Windows too
+///
+/// Basic algorithm is as follows:
+/// 
+/// Iterate through both strings
+///  If either current character is not a number, compare the rest of the strings lexically
+///  else if they're both numbers:
+///     Count the leading zeroes in both strings and skip over them
+///     Read the two numbers
+///     If they're not the same, return accordingly
+///     If they are the same, but the leading zeroes are different, return whichever has fewer zeroes as "larger"
+///     Otherwise, they're the same to continue scanning
+///  else they're both not numbers:
+///     Compare lexically till we find a number
+///
+template <typename CharType>
+int LogicalStringCompareImpl(__in const CharType* str1, __in const CharType* str2)
+{
+    Assert(str1 != nullptr);
+    Assert(str2 != nullptr);
+
+    CharType* p1 = const_cast<CharType*>(str1);
+    CharType* p2 = const_cast<CharType*>(str2);
+
+    while (*p1 && *p2)
+    {
+        bool isDigit1 = isDigit(*p1);
+        bool isDigit2 = isDigit(*p2);
+
+        // If either character is not a number, compare lexically
+        if (isDigit1 ^ isDigit2)
+        {
+            while (*p1 && *p2)
```

Comment:
Thank you @dreamer-dead - wasn't aware of lexicographical_compare, very interesting.
Generally, STL functions are ok on a case-by-case basis, STL classes/types are generally avoided for a variety of reasons (memory model inconsistencies, perf in some cases, etc). In this particular case, I would have been ok using lexicographical_compare. Looking at its definition though, couple things stand out:
- It needs the beginning and end of the sequence
- It returns a bool (so either strictly less than or not)

LogicalStringCompareImpl checks for equality, greater than or less than. Given this requirement, using lexicographical_compare would involve the following (I think- please tell me if there is a simpler way):
- Create two std::string instances from p1 and p2
- Check first for equality
- Check next with lexicographical_compare

It seems like we'd have done at least three passes at this point (once to determine the beginning/end of the string, once for the equality check and once for the lexicographical_compare. The current implementation just does a single pass.

Is there a simpler way that I'm not seeing? If not, I'd prefer to keep the current implementation.